ghcide/src/Development/IDE/Plugin/LSPWindowShowMessageRecorder.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE GADTs #-}
+
+module Development.IDE.Plugin.LSPWindowShowMessageRecorder (makeLspShowMessageRecorder) where
+
+import Control.Monad.IO.Class
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Data.Foldable (for_)
+import Data.IORef
+import Data.IORef.Extra (atomicModifyIORef'_)
+import Data.Text (Text)
+import Development.IDE.Types.Logger
+import Ide.Types (PluginDescriptor (pluginNotificationHandlers), defaultPluginDescriptor, mkPluginNotificationHandler)
+import Language.LSP.Server (LanguageContextEnv, getLspEnv)
+import qualified Language.LSP.Server as LSP
+import Language.LSP.Types (MessageType (..), SMethod (SInitialized, SWindowShowMessage), ShowMessageParams (..))
+
+-- | Creates a recorder that logs to the LSP stream via WindowShowMessage notifications.
+--   The recorder won't attempt to send messages until the LSP stream is initialized.
+makeLspShowMessageRecorder ::
+  IO (Recorder (WithPriority Text), PluginDescriptor c)
+makeLspShowMessageRecorder = do
+  envRef <- newIORef Nothing
+  -- messages logged before the LSP stream is initialized will be sent when it is
+  backLogRef <- newIORef []
+  let recorder = Recorder $ \it -> do
+        mbenv <- liftIO $ readIORef envRef
+        case mbenv of
+          Nothing -> liftIO $ atomicModifyIORef'_ backLogRef (it :)
+          Just env -> sendMsg env it
+      -- the plugin captures the language context, so it can be used to send messages
```

Comment:
I don't really like this approach. It's extremely indirect. My branch just has a function `LanguageContextEnv config -> Recorder ...` and then we have to pass it along until we reach `initialize`.

Reply:
I think we could relatively easily adapt this to just pass the `IORef` to be filled in rather than going via a plugin...

Reply:
Please send a PR! I suspect that the plugin solution is more modular, as it abstracts away the passing of the mutable state around.

Reply:
Yes, but it does so by using a mechanism that really isn't designed for that. Now the recorder mechanism is intertwined with the plugin mechanism, which it should have nothing to do with that. We shouldn't have to fix the loggers if we change how plugins work; that seems anti-modular to me.

Reply:
I'll take a look.